Fix editor widgets default behaviors.

pvalsecc opened the pull request:
When the user opens the layers properties, leaves the default
fields definition as is and saves the project, the fields were saved
with empty configuration values.

I've fixed all the editor widget types that are potentially used
without configuration.

m-kuhn commented:
Did you also consider to check for variant.isValid() in the serialization (toXml) method and skip writing such configuration options?

pvalsecc commented:
@m-kuhn, OK, I will, but I'll leave the fixes I've already added for all the project files out there that already contain empty attributes.

pvalsecc committed:
Better reading/writting of editor widget config.

Avoid saving config parameters that are not set and support not
having them in the XML.

m-kuhn commented:
I didn't look into it. According to your mail I got the feeling that empty attributes were mainly triggered by the new "smart default widgets" system which is only in master, so there shouldn't be any affected project files out there created by regular releases.

pvalsecc commented:
Yes, but doesn't cost much to help those using master and it's already done.

3nids commented:
I would avoid adding code to fix master-only projects. It is very unlikely someone rely on master for production at the moment. My 2 cts.

m-kuhn commented:
On the other hand it clutters the api?

On October 10, 2016 9:42:07 AM GMT+02:00, Patrick Valsecchi <notifications@github.com> wrote:
>Yes, but doesn't cost much to help those using master and it's already
>done.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/qgis/QGIS/pull/3585#issuecomment-252552813

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

pvalsecc commented:
OK, I've removed the "cluttering"

m-kuhn commented on src/gui/editorwidgets/core/qgseditorwidgetfactory.cpp:
```diff
@@ -123,3 +123,27 @@ unsigned int QgsEditorWidgetFactory::fieldScore( const QgsVectorLayer* vl, int f
+  if ( !value.isNull() )
+  {
+    config.insert( fieldName, value );
+  }
```
Is type handling not required in this method?

See e.g. the changes in src/gui/editorwidgets/qgswebviewwidgetfactory.cpp which have been converted toInt() before and now are stored in the runtime config as string.

pvalsecc replied:
The XML has no type info. So all the config entries are strings and all the widget support that. It was already like that before. In the config2xml I had to just check for bool to have the same format as before.

m-kuhn replied:
I think the application had knowledge of type information before, see e.g. https://github.com/qgis/QGIS/pull/3585/files#diff-22d1ee53da45d48890d2e7e5d0545e95L61 

I'm not sure if something relied on it further down the road but it's worth taking into account

pvalsecc replied:
Funny that the example you are mentioning (DocumentViewerWidth) is not used :-)

The config is QVariantMap. So storing it as a string or directly as a an int won't change much. Where it is used, we have to do the `toInt()` anyway.

I could add a `QVariant::Type` parameter to `xml2config` and call `QVariant::convert` in there, but I really don't see the point...

pvalsecc replied:
@m-kuhn, you want me to add the types or you are OK with the current?
